BUG: .reset_index() should raise with an invalid level name (GH20925)

KalyanGokhale opened the pull request:
#20925 Raises appropriate error for Series.reset_index(level_name, drop=True) when index is flat and an invalid level is supplied

- [x] closes #20925 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

KalyanGokhale committed:
GH20925

Raises appropriate error for Series.reset_index(level_name, drop=True) when index is flat and an invalid level is supplied

KalyanGokhale committed:
GH20925

Raises appropriate error for Series.reset_index(level_name, drop=True) when index is flat and an invalid level is supplied

TomAugspurger commented:
Could you add a test that the correct error and error message is raised? Use `tm.assert_raises_regex`?

KalyanGokhale committed:
GH20925

Raises appropriate error for Series.reset_index(level_name, drop=True) when index is flat and an invalid level is supplied

pep8speaks commented:
Hello @KalyanGokhale! Thanks for updating the PR.



 - In the file [`pandas/tests/series/indexing/test_indexing.py`](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py), following are the PEP8 issues :

> [Line 772:1](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L772): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 774:5](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L774): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 776:5](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L776): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 779:5](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L779): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 782:5](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L782): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 783:70](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L783): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 784:5](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L784): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 787:5](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L787): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 789:5](https://github.com/pandas-dev/pandas/blob/de734d8df231495aff5346e1c84cd2e92e2609c2/pandas/tests/series/indexing/test_indexing.py#L789): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

KalyanGokhale committed:
GH20925

Raises appropriate error for Series.reset_index(level_name, drop=True) when index is flat and an invalid level is supplied

KalyanGokhale commented:
@TomAugspurger have added the test - please let me know if any further edits are needed.

toobaz commented on pandas/tests/series/indexing/test_indexing.py:
```diff
@@ -768,3 +768,25 @@ def test_head_tail(test_data):
+    with tm.assert_raises_regex(KeyError, 'not found'):
+        s.reset_index('wrong', drop=True)
+    # Data for Test Case 4
+    s = pd.Series(range(4), name='valid')
```
This (and I mean: my [testcase](https://github.com/pandas-dev/pandas/issues/20925#issuecomment-386284093)) doesn't make much sense, as we are giving a name to ``s`` itself, not to its index. I should have written

```python
s = pd.Series(range(4), index=pd.RangeIndex(4, name='valid'))
```

Notice that if you replace with the above, the following doesn't raise any more. But it might be a bug in ``MultiIndex.droplevel``, in which case OK to postpone.

KalyanGokhale replied:
Will delete the Test Case 4 now

toobaz commented on pandas/core/series.py:
```diff
@@ -1195,12 +1195,14 @@ def reset_index(self, level=None, drop=False, name=None, inplace=False):
+            if level is not None:
                 if not isinstance(level, (tuple, list)):
                     level = [level]
+                level = set(level)
```
Why?!

KalyanGokhale replied:
`level = set(level)` is not required now. Did not remove it after combining two separate blocks post-testing. Will delete it.

toobaz commented on pandas/core/series.py:
```diff
@@ -1195,12 +1195,14 @@ def reset_index(self, level=None, drop=False, name=None, inplace=False):
-                if len(level) < len(self.index.levels):
-                    new_index = self.index.droplevel(level)
+                if isinstance(self.index, MultiIndex):
+                    if (len(level) < len(self.index.levels)):
```
``len(self.index.levels)`` can be replaced with ``self.index.nlevels`` (and then I guess you'll be able to drop the ``MultiIndex`` test above). But is this test actually needed? (It's not a rhetorical question)

KalyanGokhale replied:
Thanks for the suggestion. 

Yes replacing `len(self.index.levels)` with `self.index.nlevels` helps and `MultiIndex` test can then be removed. However, have kept the `MultiIndex` be kept to avoid any unforeseen outcomes - keeping it also produces the same outcomes (at least in the test cases checked).

`len(level) < self.index.nlevels` test is needed, in case this is dropped (assuming `MultiIndex` test is also dropped) then following errors are observed for these cases:
*Test Data*
`arrays = [np.array(['bar', 'bar', 'baz', 'baz']), np.array(['one', 'two', 'one', 'two'])]`
`s = pd.Series(range(4), name='foo', index=pd.MultiIndex.from_arrays(arrays, names=['a', 'b']))`
*Test Case*
`s.reset_index(['a', 'b'], drop=True)`
*Outcome*
`ValueError: Must pass non-zero number of levels/labels`

*Test Data*
`s = pd.Series(range(4), index=pd.RangeIndex(4, name='valid'))`
*Test Cases*
`s.reset_index(['valid', 'valid'], drop=True)`
`s.reset_index(['valid'], drop=True)`
`s.reset_index('valid', drop=True)`
*Outcome for each of 3 test cases above*
`AttributeError: 'RangeIndex' object has no attribute 'droplevel'`

toobaz replied:
> len(level) < self.index.nlevels test is needed

Right, hadn't considered that zero-levels ``MultiIndex``es can't exist.

The ``isinstance(self.index, MultiIndex)`` doesn't make much sense but it's OK to leave it, as it guarantees that if ``pd.Series(range(4)).reset_index([], drop=True)`` doesn't fail (``pd.Index.droplevel()`` doesn't exist). Could you add a test case for this?

KalyanGokhale replied:
done

toobaz reviewed (changes requested):
My test case was wrong... removing the ``set()`` will maybe fix the correct one?

KalyanGokhale committed:
GH20925

Raises appropriate error for Series.reset_index(level_name, drop=True) when index is flat and invalid level is supplied. Made edits as requested in the review.

jreback commented on pandas/tests/series/indexing/test_indexing.py:
```diff
@@ -768,3 +768,20 @@ def test_head_tail(test_data):
+    # https://github.com/pandas-dev/pandas/issues/20925
+    # Data for Test Case 1 and 2
+    s = pd.Series(range(4))
+    # Test Case 1
```
put blank lines between cases. Don't use a 'Test Case 1' as a lable, either remove of put what it is testing

KalyanGokhale committed:
GH20925

Made edits to test file for labels and introduced blank spaces between test cases.

KalyanGokhale committed:
GH20925

Made edits to test file for labels and introduced blank spaces between test cases.

KalyanGokhale committed:
GH20925

Made edits to test file for labels and introduced blank spaces between test cases.

KalyanGokhale committed:
GH20925

Made edits to test file for labels and introduced blank spaces between test cases.

toobaz commented on pandas/tests/series/indexing/test_indexing.py:
```diff
@@ -768,3 +768,19 @@ def test_head_tail(test_data):
+def test_reset_index_drop_errmsg():
+    # https://github.com/pandas-dev/pandas/issues/20925
+
+    # Check KeyError raised for series where no 'level' name is defined
```
series -> series index
no 'level' name is defined -> passed level name is missing

KalyanGokhale replied:
done

toobaz commented:
I mistakenly wrote in an obsolete discussion, copying here:

> len(level) < self.index.nlevels test is needed

Right, hadn't considered that zero-levels ``MultiIndex``es can't exist.

The ``isinstance(self.index, MultiIndex)`` doesn't make much sense but it's OK to leave it, as it guarantees that if ``pd.Series(range(4)).reset_index([], drop=True)`` doesn't fail (``pd.Index.droplevel()`` doesn't exist). Could you add a test case for this?

After that and a couple of minor fixes to comments, I think we're ready.

KalyanGokhale committed:
GH20925

made suggested edits to tests

KalyanGokhale committed:
GH20925

made suggested edits to tests

KalyanGokhale commented on pandas/tests/series/indexing/test_indexing.py:
```diff
@@ -768,3 +768,19 @@ def test_head_tail(test_data):
+    with tm.assert_raises_regex(KeyError, 'must be same as name'):
+        s.reset_index('wrong')
+
+    # Check KeyError raised for series where 'level' to be dropped is undefined
```
done

toobaz commented on pandas/tests/series/indexing/test_indexing.py:
```diff
@@ -768,3 +768,24 @@ def test_head_tail(test_data):
     assert_series_equal(test_data.series.tail(0), test_data.series[0:0])
+
+
+def test_reset_index_drop_errmsg():
```
I just noticed this is in the wrong file... ``reset_index`` is tested in ``pandas/tests/series/test_alter_axes.py``, not ``test_indexing.py``.

While you're at it, please replace ``# https://github.com/pandas-dev/pandas/issues/20925`` with the standard ``# GH 20925``.

KalyanGokhale replied:
Thanks didn't know that the right file was 'test_alter_axes.py'.

Have moved the function there (now renamed as 'test_reset_index_drop_errors' based on @TomAugspurger 's comment) - also updated comment for # GH 20925

toobaz commented on pandas/tests/series/indexing/test_indexing.py:
```diff
@@ -768,3 +768,24 @@ def test_head_tail(test_data):
+    with tm.assert_raises_regex(KeyError, 'not found'):
+        s.reset_index('wrong', drop=True)
+
+    # Check that .reset_index([],drop=True) doesn't fail
```
This is fine but probably not in the right test: could you append it to ``test_reset_index_level`` (in ``pandas/tests/series/test_alter_axes.py``)?

KalyanGokhale replied:
Done. Have moved it to function 'test_reset_index_level' outside the for loop

toobaz reviewed (changes requested):
Changes OK, just move to correct file

TomAugspurger commented on pandas/tests/series/indexing/test_indexing.py:
```diff
@@ -768,3 +768,24 @@ def test_head_tail(test_data):
     assert_series_equal(test_data.series.tail(0), test_data.series[0:0])
+
+
+def test_reset_index_drop_errmsg():
```
`errmsg` -> `errors` or `error_message`

KalyanGokhale replied:
done. renamed def as 'test_reset_index_drop_errors'

TomAugspurger reviewed (commented):
Release note can go in 0.23.1 (maybe have to merge in master for it to show up).

KalyanGokhale committed:
GH20925

Made requested changes, moved test to pandas/tests/series/test_alter_axes.py

KalyanGokhale committed:
GH20925

Made requested changes, moved test to pandas/tests/series/test_alter_axes.py

toobaz commented:
Looks all good to me, ready to merge after the whatsnew note is added.

KalyanGokhale committed:
Merge pull request #1 from pandas-dev/master

Updating to 0.23.0

KalyanGokhale committed:
Merge pull request #2 from KalyanGokhale/master

Updating to 0.23.0

KalyanGokhale committed:
Update v0.23.1.txt

KalyanGokhale commented:
Updated v0.23.1.txt under Bug Fixes > Indexing as:
*Bug in :meth:`Series.reset_index` where appropriate error was not raised with a non-named level (:issue:`20925`)*

toobaz commented on doc/source/whatsnew/v0.23.1.txt:
```diff
@@ -59,7 +59,7 @@ Conversion
 ^^^^^^^^
 
--
+- Bug in :meth:`Series.reset_index` where appropriate error was not raised with a non-named level (:issue:`20925`)
```
I don't get the "a non-named level". Wouldn't it be "an invalid level name"?

KalyanGokhale replied:
Agreed. Done.

KalyanGokhale committed:
Update v0.23.1.txt

toobaz commented:
@KalyanGokhale merged, thanks!
